### ISSUE TITLE
[Proposal] Array Ranges

### ISSUE THREAD

**Issue by VBAndCs** [dotnet/vblang#559]
I suggest to allow:
`Dim A = {1 To 5}`
which means:
`Dim A = {1, 2, 3, 4, 5}`
we can use such ranges in copying arrays too:
`Dim B = A[2 To 4]`
B will have `{3, 4, 5}` since index 2 refers to the third element in the array. In fact 0-based indexing will be a bit confusing here, and we can argue if we should use 1-based indexing in this particular syntax or not!
This cab also be used with strings:
```
Dim x = "This is a test"
Console.WriteLine(x[10 to 13]) ' Test
```

**Comment by zspitz** [comment]
The purported 0-based indexing confusion is only because the values in this particular array begin with `1`.

**Comment by pricerc** [comment]
I have sympathy with:
> I suggest to allow:
> `Dim A = {1 To 5}`
> which means:
> `Dim A = {1, 2, 3, 4, 5}`

But the rest will just make code harder to understand.

We already have substring and array copy functions in the .NET libraries which are quite clear in their intent and semantics, and are not particularly verbose.

@VBAndCs probably knows my feelings on this quite well by now, but one of VB's greatest strengths is how easy it is to read, even for people who are not programmers. 

In my opinion, this is abbreviating code for the sake of abbreviating code, and does not help the language.

**Comment by VBAndCs** [comment]
What is not readable about: x of 10 to 13? This is exactly what `x[10 to 13]`. You can ask non programmers about its meaning and see how many can get it.

**Comment by VBAndCs** [comment]
Also, having this as a lang feature, will be shorter and cleaner, and allow VB to do optimizations like using spans when possible. It also allows to regain an old vb6 feature:
`X[ 10 to 13] = "example"`
So, it can act as a replace expression too, for both lists and strings.
I know string is immutable, but I see no harm in allowing:
`X(1) = "some string"`
which can be lowered to:
```
x = x.Substring(0) + "some string" + x.Substring(2)
```

VB.NET need to be smart again, to have a future.

**Comment by VBAndCs** [comment]
Note that this way of thinking:
> The frame work has a solution

is combined with this way of thinking:

> C# can do it all.

so, we are here:

> No need for VB.NET anymore!

So, please forget about the class library and C# way of thinking, and lets regain VB spirit that made it dominate for decades: a smart, short and readable language.

**Comment by pricerc** [comment]
> What is not readable about: x of 10 to 13? This is exactly what `x[10 to 13]`. You can ask non programmers about its meaning and see how many can get it.

With respect, it doesn't even make sense to me.

**Comment by pricerc** [comment]
> It also allows to regain an old vb6 feature:
> `X[ 10 to 13] = "example"`

?
That's not a VB6 syntax that I recognize. And I know it won't compile under VBA.

**Comment by pricerc** [comment]
> So, it can act as a replace expression too, for both lists and strings.
> I know string is immutable, but I see no harm in allowing:
> `X(1) = "some string"`
> which can be lowered to:

``` VB
    Dim x As String
    x = "Hello, World! How are you?"
    
    mid(x,10,4) = "example"
    
    Debug.Print (x)
 ```

works today.

**Comment by pricerc** [comment]
> So, please forget about the class library and C# way of thinking

Respectfully; I disagree. 

A "Class library way of thinking" and a "C# way of thinking" are two very different things with important differences, and combining them is unhelpful.

Object-oriented programming is all about the "Class library way of thinking". Why would anyone want to do again what someone has already done? The only reason for doing that would because of "Not invented here" syndrome, which is also not helpful.

Trying to add a new syntax to do something in VB that is already easily accessed through a library function does not seem to me to be a good use of compiler-writer resources.

> Note that this way of thinking:
> 
> > The frame work has a solution
> 
> is combined with this way of thinking:
> 
> > C# can do it all.
> 
> so, we are here:
> 
> > No need for VB.NET anymore!
> 

I don't see it that way at all.

C# and VB are two different ways of using the same class libraries. That's it. 

Neither can "do it all".

And both have good and bad aspects.

For me, a bad aspect for C# is too many symbols and not enough words. Your last few proposals for VB have been about using symbols to replace words that we already have, and so make VB look more like C#. 

And so I don't think they would help the language grow, or solve any significant problem.

**Comment by VBAndCs** [comment]
Yes I was refering to mid, but it preservs the original length, which has no purpose since strings in .NET are immutable, and mid will create a new instance anyway. So:
```
        Dim x = "abcd"
        mid(x, 1, 2) = "example "
        Console.WriteLine(x) ' excd
```

This is why I didn't use mid statement for 2 decades, since it is rarely useful.
So, I need:
```
        Dim x = "abcd"
        x(0 to 1) = "example "
        Console.WriteLine(x) ' example cd
```

**Comment by VBAndCs** [comment]
> I don't see it that way at all.

Microsoft does, so, here we are.

**Comment by pricerc** [comment]
> This is why I didn't use mid statement for 2 decades, since it is rarely useful.
> So, I need:
> 
> ```
>         Dim x = "abcd"
>         x(0 to 1) = "example "
>         Console.WriteLine(x) ' example cd
> ```

I can see the usefulness in that, but I'm not comfortable with the syntax.

For me, *good* VB syntax is immediately obvious without needing to resort to instructions to figure out what something means. I don't get that from this syntax.
